wp-includes/mysql/class-wp-mysql-lexer.php

```diff
@@ -0,0 +1,2898 @@
+	 * the tokenization process, recognizing all digit-only sequences as numbers.
+	 *
+	 * Rules:
+	 *   1. Allowed characters are ASCII a-z, A-Z, 0-9, _, $, and Unicode U+0080-U+FFFF.
+	 *   2. Unquoted identifiers may begin with a digit but may not consist solely of digits.
+	 *
+	 *  See:
+	 *    https://dev.mysql.com/doc/refman/8.4/en/identifiers.html
+	 */
+	private function read_identifier(): int {
+		$started_at = $this->bytes_already_read;
+		while ( true ) {
+			// First, let's try to parse an ASCII sequence.
+			$this->bytes_already_read += strspn(
+				$this->sql,
+				'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_$',
+				$this->bytes_already_read
+			);
+
+			// Check if the following byte can be part of a multibyte character
+			// in the range of U+0080 to U+FFFF before looking at further bytes.
+			// If it can't, bail out early to avoid unnecessary UTF-8 decoding.
+			// Identifiers are usually ASCII-only, so we can optimize for that.
+			$byte_1 = ord(
+				$this->sql[ $this->bytes_already_read ] ?? ''
+			);
+			if ( $byte_1 < 0xC2 || $byte_1 > 0xEF ) {
+				break;
+			}
+
+			// Look for a valid 2-byte UTF-8 symbol. Covers range U+0080 - U+07FF.
+			$byte_2 = ord(
+				$this->sql[ $this->bytes_already_read + 1 ] ?? ''
+			);
+			if (
+				$byte_1 <= 0xDF
+				&& $byte_2 >= 0x80 && $byte_2 <= 0xBF
+			) {
+				$this->bytes_already_read += 2;
+				continue;
+			}
+
+			// Look for a valid 3-byte UTF-8 symbol in range U+0800 - U+FFFF.
+			$byte_3 = ord(
+				$this->sql[ $this->bytes_already_read + 2 ] ?? ''
+			);
+			if (
+				$byte_1 <= 0xEF
+				&& $byte_2 >= 0x80 && $byte_2 <= 0xBF
+				&& $byte_3 >= 0x80 && $byte_3 <= 0xBF
+				// Exclude surrogate range U+D800 to U+DFFF:
+				&& ! ( 0xED === $byte_1 && $byte_2 >= 0xA0 )
+				// Exclude overlong encodings:
+				&& ! ( 0xE0 === $byte_1 && $byte_2 < 0xA0 )
+			) {
+				$this->bytes_already_read += 3;
+				continue;
+			}
+
+			// Not a valid identifier character.
+			break;
+		}
+
+		// An identifier cannot consist solely of digits, but we don't need to
+		// ensure that explicitly, as numbers are processed before identifiers.
+
+		return $this->bytes_already_read - $started_at > 0
+			? self::IDENTIFIER
+			: self::INVALID_INPUT;
+	}
+
+	private function read_number(): int {
+		// @TODO: Support numeric-only identifier partes after "." (e.g., 1ea10.1).
+
+		$byte      = $this->sql[ $this->bytes_already_read ] ?? null;
+		$next_byte = $this->sql[ $this->bytes_already_read + 1 ] ?? null;
+
+		if (
+			// HEX number in the form of 0xN.
+			( '0' === $byte && 'x' === $next_byte )
+			// HEX number in the form of x'N' or X'N'.
+			|| ( ( 'x' === $byte || 'X' === $byte ) && "'" === $next_byte )
+		) {
+			$is_quoted                 = "'" === $next_byte;
+			$this->bytes_already_read += 2; // Consume "0x" or "x'".
+			$this->bytes_already_read += strspn( $this->sql, '0123456789abcdefABCDEF', $this->bytes_already_read );
```

Comment:
What if that `strspn` returns 0? Should `0xz` or `0xEOF` be a syntax error / paused state? Ditto for all other `strspn` calls. Maybe we don't need to solve it now and a `@TODO` comment would be enough.